west: Introduce run once commands (e.g. erase, recover) and deferred reset

nordicjm opened the pull request:
Introduces additional fields to soc.yml or board.yml which can be used to specify `west flash` commands that should be grouped for multiple boards when using sysbuild, this allows usage of e.g. `--erase` to allow the flash to be erased once when flashing a set of images rather than once per image which would erase what had just been programmed. The format of this is the command then a list of boards which are part of the group whereby it will only be executed one for. Also supported is grouping boards to prevent resetting of the CPUs after programming when multiple images are being programmed until all images for the grouped cores have been programmed, at which point the CPU can be reset, this is ideal for multiple cores or things like TF-M.

TODO list:

- [x] documentation

nordicjm committed:
west: runners: Add run once commands and deferred reset

This adds supports for flashing images with sysbuild where there
are multiple images per board to prevent using the same command per
image flash which might cause issues if they are not ran just once
per flash per unique board name. A deferred reset feature is also
introduced that prevents a board (or multiple) from being reset if
multiple images are to be flashed until the final one has been
flashed which prevents issues with e.g. security bits being enabled
that then prevent flashing further images.

These options can be set at a board level (in board.yml) or a SoC
level (in soc.yml), if both are present then the board configuration
will be used instead of the SoC, and regex can be used for matching
of partial names which allows for matching specific SoCs or CPU cores
regardless of the board being used

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
soc: nordic: Add run once and deferred reset configuration

Adds configuration that allows nRF53 and nRF91-based boards to be
flashed through west using sysbuild for multiple images with the
recover or erase options and prevent running those commands for
each image being flash, which would make the device unbootable.
Also defers reset whilst all images for the cores of these SoCs
are flashed.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented on scripts/schemas/board-schema.yml:
```diff
@@ -78,3 +78,56 @@ mapping:
     type: seq
     sequence:
       - include: board-schema
+  single_commands:
```
Fixed

nordicjm commented on scripts/schemas/board-schema.yml:
```diff
@@ -78,3 +78,56 @@ mapping:
+                              A list of board names (or regex) which this should apply to. Regex
+                              can be used to match a specific SoC or CPU but must only match one
+                              for a given board.
+  delay_reset:
```
Fixed

carlescufi commented on soc/nordic/soc.yml:
```diff
@@ -1,39 +1,73 @@
+run_once:
+  - '--recover':
+      - boards:
+          - /(.+)/nrf5340/cpunet/
```
Why would you have an entry with a single board target like this? what's the point of saying "run once" for a single target?

carlescufi replied:
OK, I see!

gmarull commented on soc/nordic/soc.yml:
```diff
@@ -1,39 +1,73 @@
+          - name: nrf9161
+
+# Recovery/erase is only needed once per core.
+run_once:
```
a top-level "run_conce" key feels weird here, I'd suggest something like:

```
- flashing:
  - singlerun:
  ...
  - anyfuturestuff:  
```

gmarull replied:
(or `flash.yml`...)

nordicjm replied:
This was a separate file initially but moved into these files because "of too many yaml files" concerns raised by others

gmarull replied:
> This was a separate file initially but moved into these files because "of too many yaml files" concerns raised by others

IMHO mixing files purpose is not a good idea either.

tejlmand replied:
> IMHO mixing files purpose is not a good idea either.

feel free to join the discussion: https://github.com/zephyrproject-rtos/zephyr/pull/50305#discussion_r1400797335
I feel the same, multiple files can be easier to read understand if the are well named and organized, quote:
> The fact that both files are in yaml doesn't mean there should be one huge yaml file containing everything.

But there was request to unify and have a single source for various meta data:
https://github.com/zephyrproject-rtos/zephyr/pull/50305#discussion_r1407873481

gmarull commented on soc/nordic/soc.yml:
```diff
@@ -1,39 +1,73 @@
+# for each core have been flashed, this allows security bits to be set during
+# programming without them interfering with additional flashing processes.
+deferred_reset:
+  - boards:
+      - /(.+)/nrf5340/cpuapp/
+      - /(.+)/nrf5340/cpuapp/ns/
+  - boards:
+      - /(.+)/nrf5340/cpunet/
+  - boards:
+      - /(.+)/nrf9160/
+      - /(.+)/nrf9160/ns/
```
why N `boards`? can't they be merged into a single boards list?

nordicjm replied:
No, they are each groups. If you imagine a board with an nrf5340 and nrf9160 on it then as a single group it would mean one of those would be reset and the other would not

gmarull replied:
I think we need to make this more intuitive

nordicjm replied:
I don't see how that can be done, there needs to be a grouping, you can't know this implicitly for all socs as it's going to vary between different manufacturers

tejlmand replied:
then give it a dedicated group field / name.
Right now, it's not intuitive at all that two `boards:` lists actually means two groups.

And explain what a group means.

I have very similar comment which I wrote before seeing this comment.

tejlmand commented on scripts/schemas/board-schema.yml:
```diff
@@ -78,3 +78,56 @@ mapping:
+    sequence:
+      - type: map
+        mapping:
+          regex;(.*):
```
this makes the command to become a key in the yaml file, not a value.

As I see it, there are two main implecations of such approach
- The free form `.*` means there is effectively no schema check, as anything is allowed for a value.
- A `:` must suddenly follow the `value`.
   In yaml, a value is either after the colon, like:
   `<key>: <value>`
   or a list of values, like:
   ```
   my_list:
   - value1
   - value2
   - value3
   - ...
   ```

but you're creating a list of maps, but where the `key` in that list is actually the value.

I don't see why regular `key: value` mapping cannot be used, and thereby avoid tricks like this.

nordicjm replied:
Unification with snippets

tejlmand replied:
but not unified with boards.yml.

For example, the board.yml for nrf9160dk does:
```
board:
  name: nrf9160dk
  vendor: nordic
  socs:
  - name: nrf9160
    variants:
    - name: 'ns'
  - name: nrf52840
```

and **not**:
```
board:
  name: nrf9160dk
  vendor: nordic
  socs:
  - nrf9160:
      variants:
      - name: 'ns'
  - nrf52840:
```

The fact that snippet does this does not justify this principle in board or soc yaml files.

Especially not when this proposal introduces inconsistency on how things are done within the same yaml file.

tejlmand replied:
See also comment here: https://github.com/zephyrproject-rtos/zephyr/pull/69748#discussion_r1530031405

tejlmand commented on scripts/schemas/board-schema.yml:
```diff
@@ -78,3 +78,56 @@ mapping:
+                    required: true
+                    type: seq
+                    desc: |
+                          A grouping of boards which the command should apply to. Can be used
+                          multiple times to have multiple groups.
```
what does a group mean in this context ?

Why not just a list of boards ?
What happens if I move a board from one group to another, does it change anything ?

nordicjm replied:
It changes a lot, see https://github.com/zephyrproject-rtos/zephyr/pull/69748#discussion_r1517464462

tejlmand commented on scripts/schemas/board-schema.yml:
```diff
@@ -78,3 +78,56 @@ mapping:
+                    desc: |
+                          A grouping of boards which the command should apply to. Can be used
+                          multiple times to have multiple groups.
+                    sequence:
+                      - type: str
```
This seems to allow users to refer to any board, regardless of this board being described in this yaml or not.

For example:
```
board:
  name: plank
  socs:
  - name: soc1
run_once:
  - '--a-command':
     - boards:
        - 'foo'
```

It should not be possible to reference boards beside those defined by this board.yml.

nordicjm replied:
We have a huge problem with `west boards` being incredibly slow because of schema validation, and to add extra overhead on top of this of then doing regex matching every time it is invoked does not really make sense. Verification as part of compliance in CI makes sense but whenever someone runs `west flash` no

tejlmand replied:
>  and to add extra overhead on top of this of then doing regex

are your really arguing that a regex like this is faster ?
```
        mapping:
          regex;(.*):
```

then I would like to see some performance numbers.

tejlmand replied:
I still lack some performance numbers.

Also, this principle has high risk of wrongly indenting yaml code, for example:
```
    - '--recover':
      run: first
      runners:
      - nrfjprog
      - jlink
```
is wrong, but this works:
```
    - '--recover':
            run: first
            runners:
            - nrfjprog
            - jlink
```
This minor detail of remembering the extra indentation is extremely important but may not be obvious to everyone, and the schema validation failure is not helpful at all.

Thus, I still argue that a normal key-value pair is much better, in this case like this:
```
    - command: '--recover':
      run: first
      runners:
      - nrfjprog
      - jlink
```

No surprises to users on how to do indentation.

tejlmand replied:
as for performance, the first reason for `west boards` is the split in almost 500 boards.yml files, which all need to be validated, not whether you try to sub-optimize the key.

In fact, the overhead of `west` itself is quite significant.
Running `west boards --board foo` gives: (`--board foo` is to avoid time overhead of printing)
```
$ time west boards --board foo
_______________________________________________________
Executed in  715.04 millis    fish           external
   usr time  698.32 millis    0.00 micros  698.32 millis
   sys time   16.36 millis  304.00 micros   16.05 millis
```

but removing `west` and use `list_boards.py` directly, give you:
```
time ./scripts/list_boards.py --board-root=(pwd) --arch-root=(pwd) --soc-root=(pwd) --board=foo
________________________________________________________
Executed in  543.97 millis    fish           external
   usr time  514.78 millis    0.00 micros  514.78 millis
   sys time   28.28 millis  350.00 micros   27.93 millis
```

nordicjm replied:
> Also, this principle of using values for keys through the regex has high risk of users wrongly indenting yaml code, for example:
> This minor detail of remembering the extra indentation is extremely important but may not be obvious to everyone, and the schema validation failure is not helpful at all.

The user won't get into that situation because the schema would fail

tejlmand replied:
> The user won't get into that situation because the schema would fail

any user creating a oot board has the risk of falling into this.

And any Zephyr contributor can waste a lot of time on this.

nordicjm replied:
Then I suggest you improve the error reporting by submitting a PR to https://pypi.org/project/pykwalify/

tejlmand commented on scripts/schemas/board-schema.yml:
```diff
@@ -78,3 +78,56 @@ mapping:
+            desc: |
+                  A list of commands which should be limited to running once per invocation of
+                  west flash for a set of boards
+            sequence:
+              - type: map
+                mapping:
+                  boards:
```
Why is there a list `boards` with a list of board names ?
What I refer to is, why this:
```
<key>:
- boards:
  - foo
  - bar
- boards:
  - 
```
Instead of:
```
command: <value>
boards:
- foo
- bar
- ...
```

tejlmand commented on scripts/schemas/soc-schema.yml:
```diff
@@ -70,3 +70,56 @@ mapping:
+                  west flash for a set of boards
+            sequence:
+              - type: map
+                mapping:
+                  boards:
```
soc.yml should not refer to boards.

Please change this to socs, so that soc.yml just define the socs to which the command belongs, then the logic in Python must map this to the relevant boards.

Just like what we do with CPU clusters. The soc.yml doesn't define the boards to which the CPU clusters applies, instead the board inherits the CPU clusters defined by the SoC the board is using.

Similar, let the SoC define the commands and the board using that SoC can then inherit the commands.

nordicjm replied:
This would create a mass of redundant extra code to perform lookups without any actual gain

nordicjm replied:
> Similar, let the SoC define the commands and the board using that SoC can then inherit the commands.

As discussed and decided at architecture working group this is purposely not the cause, a board will override the whole soc configuration

tejlmand commented on scripts/west_commands/run_common.py:
```diff
@@ -165,21 +198,116 @@ def do_run_common(command, user_args, user_runner_args, domains=None):
+                # and one image in total.
+                board_image_count[board] = ImagesFlashed(0, 1)
+
+            # Load board flash runner configuration (if it exists) and store
+            # single-use commands in a dictionary so that they get executed
+            # once per unique board name.
```
unique board name or unique board dir ?

There is quite a difference between those two, as a board.yml may define multiple unique boards.

nordicjm replied:
Board name as the text says

tejlmand commented on scripts/west_commands/run_common.py:
```diff
@@ -165,21 +198,116 @@ def do_run_common(command, user_args, user_runner_args, domains=None):
+                except FileNotFoundError:
+                    continue
+
+                if 'run_once' not in data_yaml and 'deferred_reset' not in data_yaml:
+                    # Check SoC file
```
why do we only check SoC file if both of those are not present in the board.yml ?

To me it seems that board.yml could be defining commands only for some board targets, while others should be inheriting commands from the soc.yml.

tejlmand commented on scripts/west_commands/run_common.py:
```diff
@@ -165,21 +198,116 @@ def do_run_common(command, user_args, user_runner_args, domains=None):
+                        with open(soc_yaml_file, 'r') as f:
+                            data_yaml = yaml.safe_load(f.read())
+
+                    except FileNotFoundError:
+                        continue
```
something to handle ?

nordicjm replied:
No, hwmv1 so can be ignored

tejlmand replied:
then please add a comment in code, for example:
```suggestion
                        # Legacy boards don't have a board.yml or soc.yml.
                        continue
```

tejlmand commented on soc/nordic/soc.yml:
```diff
@@ -1,39 +1,73 @@
+          - /(.+)/nrf5340/cpunet/
+      - boards:
+          - /(.+)/nrf5340/cpuapp/
+          - /(.+)/nrf5340/cpuapp/ns/
```
A soc.yml should really not know anything about board variants.

A board variant is defined by the board and could be named anything.

nordicjm replied:
That's up to vendors if they want to change, and if they do they can add this into their board.yml file, it is not an issue

tejlmand reviewed (changes requested):
Main issues I see with current approach relates to yaml file / schema:

- Multiple `boards:` entries seems to be creating a group, but it's not clear what a group is, and whether such a group should have a name. Should be made clearer.
- Misuse of yaml keys, so that instead of being a key, it actually is a value.
   Please use a `key: value` pair for such cases.
- `board.yml` seems to be able to reference board names which have no relation to the boards defined by the `board.yml`. Please ensure that only boards defined by the board.yml itself can be referenced in the commands sections. Similar for the `soc.yml`.

carlescufi commented:
Architecture WG:

- @nordicjm explains the feature
- @henrikbrixandersen asks if this can be overridden from a board. The answer is yes, you can override whatever the default is for the SoC
- @nashif asks for other vendors to take a look so that we can ensure that the `soc.yml` doesn't get out of hand if we keep inserting
- @henrikbrixandersen suggests having this information in a snippet's yml file
- @nashif wants to generalize this so that it is applicable to other vendors' use cases

nordicjm commented:
@nashif Have updated the schema, please check this is what you had in mind
@tejlmand Please re-review

carlescufi reviewed (dismissed):
This looks great to me as it is now, much better than the original.

57300 commented on scripts/west_commands/run_common.py:
```diff
@@ -165,21 +194,113 @@ def do_run_common(command, user_args, user_runner_args, domains=None):
+        # parameters.
+        board_names = []
+        for d in domains:
+            if d.build_dir is None:
+                build_dir = get_build_dir(user_args)
+            else:
+                build_dir = d.build_dir
```
nit: I don't think `d.build_dir` can be None, provided that the `Domains` class is initialized properly.

nordicjm replied:
A user can use `west flash` without being in a build directory

57300 commented on scripts/west_commands/run_common.py:
```diff
@@ -165,21 +194,113 @@ def do_run_common(command, user_args, user_runner_args, domains=None):
+            cache = load_cmake_cache(build_dir, user_args)
+            board = cache['CACHED_BOARD']
+
+            if board not in board_names:
+                board_names.append(board)
```
Looks like `board_names` should be a `set()`

nordicjm replied:
Fixed

57300 commented on scripts/west_commands/run_common.py:
```diff
@@ -165,21 +194,113 @@ def do_run_common(command, user_args, user_runner_args, domains=None):
+            if board not in board_names:
+                board_names.append(board)
+
+            if board in board_image_count:
+                board_image_count[board].total = board_image_count[board].total + 1
+            else:
+                # This sets up the default entry for a board, no flashed images
+                # and one image in total.
+                board_image_count[board] = ImagesFlashed(0, 1)
```
You could avoid this construction by initializing `board_image_count = defaultdict(ImagesFlashed)`
That should let you reduce this to:
```suggestion
            board_image_count[board].total += 1
```

nordicjm replied:
Fixed

57300 commented on scripts/west_commands/run_common.py:
```diff
@@ -165,21 +194,113 @@ def do_run_common(command, user_args, user_runner_args, domains=None):
+                group_type = 'boards'
+
+                # Search for flash runner configuration, board takes priority over SoC
+                try:
+                    with open(board_yaml_file, 'r') as f:
+                        data_yaml = yaml.safe_load(f.read())
+
+                except FileNotFoundError:
+                    continue
```
You could simplify by taking advantage of pathlib:
```suggestion
                if not board_yaml_file.is_file()
                    continue
                data_yaml = yaml.safe_load(board_yaml_file.read_text())
```

nordicjm replied:
That changes the code from loading from the board yaml file then soc yaml if that fails to just skipping everything if the board yaml file does not exist. The board file should exist so perhaps the error should be changed but the current flow is okay

57300 replied:
> loading from the board yaml file then soc yaml if that fails

Fails in what way? You're not catching a YAML parsing exception here, for example. You're moving on to `soc.yml` only if `board.yml` doesn't contain a `runners` key.

> skipping everything if the board yaml file does not exist

That's what the existing code does when catching `FileNotFoundError`.
But feel free to dismiss the suggestion, I guess it's more of a nit.

57300 commented on scripts/west_commands/run_common.py:
```diff
@@ -165,21 +194,113 @@ def do_run_common(command, user_args, user_runner_args, domains=None):
+                    except FileNotFoundError:
+                        continue
+
+                processed_boards[cache['BOARD_DIR']] = True
```
Looks like `processed_boards` should be a `set()`

nordicjm replied:
Fixed

57300 commented on scripts/west_commands/run_common.py:
```diff
@@ -165,21 +194,113 @@ def do_run_common(command, user_args, user_runner_args, domains=None):
+                # tracked
+                for i, entry in enumerate(used_cmds):
+                    for l, match in enumerate(entry.boards):
+                        if match[0:1] == '/' and match[:1] == '/' and \
```
Was this meant to be:
```suggestion
                        if match[0] == '/' and match[-1] == '/' and \
```

nordicjm replied:
Fixed

57300 commented on scripts/west_commands/run_common.py:
```diff
@@ -165,21 +194,113 @@ def do_run_common(command, user_args, user_runner_args, domains=None):
+                for i, entry in enumerate(used_cmds):
+                    for l, match in enumerate(entry.boards):
+                        if match[0:1] == '/' and match[:1] == '/' and \
+                           re.match(match[1:len(match)-1], board) is not None:
```
If you want to align the regex usage with snippets, then this should be a full match:
```suggestion
                           re.match(fr'^{match[1:-1]}$', board) is not None:
```

nordicjm replied:
Fixed

57300 commented on scripts/west_commands/run_common.py:
```diff
@@ -165,21 +194,113 @@ def do_run_common(command, user_args, user_runner_args, domains=None):
+                    for l, match in enumerate(entry.boards):
+                        if match[0:1] == '/' and match[:1] == '/' and \
+                           re.match(match[1:len(match)-1], board) is not None:
+                            used_cmds[i].boards[l] = board
```
Why modify `used_cmds` in place?

Also, this won't work if a regex matches multiple board targets. For example, I tried to simplify one of the `--recover` groups to `/(.+)/nrf5340/(.+)/` yet the operation was executed twice.

nordicjm replied:
> Also, this won't work if a regex matches multiple board targets. For example, I tried to simplify one of the --recover groups to /(.+)/nrf5340/(.+)/ yet the operation was executed twice.

Correct, it is not meant to

57300 replied:
> > Also, this won't work if a regex matches multiple board targets. For example, I tried to simplify one of the --recover groups to /(.+)/nrf5340/(.+)/ yet the operation was executed twice.
> 
> Correct, it is not meant to

That's unexpected. A regex should be able to match multiple board targets in a group. Otherwise, it deviates from snippets and becomes more difficult to document and understand.

nordicjm replied:
Now that the regex does a full match instead of partial I will look into this

nordicjm replied:
Has been reworked and implemented

57300 commented on scripts/west_commands/run_common.py:
```diff
@@ -165,21 +194,113 @@ def do_run_common(command, user_args, user_runner_args, domains=None):
+        if len(used_cmds[i].boards) == 0:
+            del used_cmds[i]
+
+        i = i - 1
```
These loops aren't very Pythonic. Use list comprehensions where possible.

nordicjm replied:
Has been reworked

57300 commented on scripts/west_commands/run_common.py:
```diff
@@ -165,21 +194,113 @@ def do_run_common(command, user_args, user_runner_args, domains=None):
 
-def do_run_common_image(command, user_args, user_runner_args, build_dir=None):
+def do_run_common_image(command, user_args, user_runner_args, build_dir=None,
+                        used_cmds=None, board_image_count=None):
```
These arguments can't really be None, so they should be mandatory.

nordicjm replied:
used_cmds, board_image_count):
    ^^^^^^^^^
SyntaxError: non-default argument follows default argument

57300 replied:
Right, you can move them before `board_dir`

57300 commented on scripts/west_commands/run_common.py:
```diff
@@ -165,21 +194,113 @@ def do_run_common(command, user_args, user_runner_args, domains=None):
     board = cache['CACHED_BOARD']
 
+    if board_image_count is not None and board in board_image_count:
+        board_image_count[board].flashed = board_image_count[board].flashed + 1
```
+=

nordicjm replied:
Fixed

57300 commented on scripts/west_commands/run_common.py:
```diff
@@ -201,6 +322,84 @@ def do_run_common_image(command, user_args, user_runner_args, build_dir=None):
+    if len(used_cmds) > 0 and len(runner_args) > 0:
+        l = len(runner_args) - 1
+        while l >= 0:
+            i = 0
+            while i < len(used_cmds):
```
`i` is not needed.
```suggestion
            for cmd in used_cmds:
```

nordicjm replied:
Fixed

57300 commented on scripts/west_commands/run_common.py:
```diff
@@ -201,6 +322,84 @@ def do_run_common_image(command, user_args, user_runner_args, build_dir=None):
+                if used_cmds[i].command == runner_args[l] and (runner_name in used_cmds[i].runners \
+                                                               or 'all' in used_cmds[i].runners):
+                    # Check if board is here
+                    found = False
+                    for check in used_cmds[i].boards:
+                        if board == check:
+                            found = True
+                            break
```
This is checking: `board in used_cmds[i].boards`

nordicjm replied:
Fixed

57300 commented on scripts/west_commands/run_common.py:
```diff
@@ -201,6 +322,84 @@ def do_run_common_image(command, user_args, user_runner_args, build_dir=None):
+                            found = True
+                            break
+
+                    if found is True:
```
Don't use `is` with booleans and other literals.

nordicjm replied:
Fixed

57300 commented on scripts/west_commands/run_common.py:
```diff
@@ -201,6 +322,84 @@ def do_run_common_image(command, user_args, user_runner_args, build_dir=None):
+                        if used_cmds[i].first is False:
+                            # For last run instances, we need to check that this really is the last
+                            # image of all boards being flashed
+                            for board in used_cmds[i].boards:
+                                if board_image_count[board].flashed != board_image_count[board].total:
+                                    found = False
+                                    break
```
This overwrites `board` so it's no longer equal to `cache['CACHED_BOARD']` in the function's scope.
It's safer to use a generator instead, for example:
```suggestion
                            found = all(
                                board_image_count[board].flashed == board_image_count[board].total
                                for board in used_cmds[i].boards
                            )
```

nordicjm replied:
Has been reworked

57300 commented on soc/nordic/soc.yml:
```diff
@@ -1,39 +1,98 @@
-    - name: nrf9151
-    - name: nrf9160
-    - name: nrf9161
+  - name: nordic_nrf
```
Is this reindentation necessary? If so, can it be done in a separate commit?

nordicjm replied:
Added

tejlmand replied:
it's not needed, please remove this extra indentation.

If we want extra indentation for yaml sequences, then I think that should be brought up as a coding guideline, else we can keep changing formatting due to personal preferences.

nordicjm committed:
soc: nordic: Reformat soc.yml file

Reformats the soc.yml file to have uniform 2-space indentation

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

tejlmand commented on soc/nordic/soc.yml:
```diff
@@ -37,3 +37,62 @@ family:
+              - /(.+)/nrf52([0-9]{3})((.+)|$)/
+          - qualifiers:
+              - /(.+)/nrf5340/cpunet/
+          - qualifiers:
+              - /(.+)/nrf5340/cpuapp/
+              - /(.+)/nrf5340/cpuapp/ns/
```
qualifier is a board terminology, please refer to SoC or CPU cluster.

Also as mentioned in last review, then we should still not use board information inside soc.yml but limit strictly to SoC and CPU cluster.
That is no `(.+)` to indicate board name.

The use of slash and different meanings within the same context is problematic.
The first and last slash is means something very different than the slashes used inside the match string.

Normally when same character is used for two different purposes, and you need to match a character literally then you would need to escape the character, in this case as: `\/`.

ref: https://github.com/zephyrproject-rtos/zephyr/blob/10f1071802ce8735ac022c4a7cfa46bf1cfaed8f/scripts/west_commands/run_common.py#L263

nordicjm replied:
Normally you would but these are stripped by the python code, so it is the same format as snippets. They are a board terminology but this is how they are mapped, it doesn't make any sense to do mapping with a key for soc or cpu cluster, and those cannot apply to the non-secure targets or other common variants

nordicjm replied:
`\/` looks ugly and would need to be stripped by the code, I don't see any compelling reason to do that. An alternative character could be used but that would then break snippets and need to be adapted and I also don't see a compelling reason to do that. It can be documented that the name starts and ends with `/` and that these are stripped so do not need escaping somewhere, might already be for snippets

nordicjm replied:
Seems snippets actually uses cmake for regex and not python

57300 replied:
I'm not in favor of `\/` either. I think it only makes sense for something like JavaScript, where regexes are part of the language, so something like this wouldn't fly:
```js
/(.+)/nrf5340/cpuapp/ /* Uncaught SyntaxError: Invalid regular expression flags */
```
For this PR, I don't mind keeping `/.../` as is, but we could consider some alternatives:
* Have different key names for when a value is a regex or not
* Always use regex implicitly

Whatever is chosen here should also be applied for snippets, so it shouldn't be a blocker anyway.

tejlmand replied:
why not take the obvious solution.
A regex matching the whole board from start to end would look like:
`^foo/bar$`, `^(.+)/bar$`.
Any literal `^` or `$` would anyway need to be escaped today in order to work correctly.

It would also mean that code like this:
https://github.com/zephyrproject-rtos/zephyr/blob/4372c1b6c0eab652151842fc105d04094bd86224/scripts/west_commands/run_common.py#L264

wouldn't have to explicitly add `^` and `$` around `{match}`

This follows regular regex matching and makes it clear in the yaml that it is actually matching from start to end, and avoid hiding this knowledge in the python code.

Strings without `^` and `$` is then treated as fixed board names.
No reason to invent our custom `/<search-string>/` pattern.

In fact, any command / program which indicates a search using `/` needs to have following `/`s escaped.
Using `sed` to replace `/foo` with `/bar` is `sed 's/\/foo/\/bar/g'`
Searching for `foo/bar` in vim, one must use `/foo\/bar`.

tejlmand replied:
> That would cause even more confusion. People would expect those strings to be regexes too.

I don't think more confusing than encapsulating all `/<text>/` pattern as regexes within `^` and `$` inside Python.

But then the only unambiguous solution would be to explicit state when a regex in the list is used, like this:
```
          - qualifiers:
            - name: foo/bar
            - regex: ^/(.+)/nrf5340/cpuapp/ns/$
```

because any other character, like `|`, `?`, etc all have a special meaning in regex.

tejlmand replied:
> Always use regex implicitly

Would of course also be an option, and considering that this code only iterates over a very limited set of boards, then I don't think we should actually worry about performance.

tejlmand reviewed (changes requested):
Still don't see the value in using regex'es for keys inside yaml.

Especially not when also consider how this principle increases the risk of making a wrongly formatted yaml file.
Ref: https://github.com/zephyrproject-rtos/zephyr/pull/69748#discussion_r1530031405 and https://github.com/zephyrproject-rtos/zephyr/pull/69748#discussion_r1519546394

nordicjm commented on scripts/schemas/board-schema.yml:
```diff
@@ -78,3 +78,56 @@ mapping:
+  runners:
+    type: map
+    mapping:
+      run_once:
```
Because additional config might be added in future that has no relation to running once

nordicjm commented on scripts/schemas/board-schema.yml:
```diff
@@ -78,3 +78,56 @@ mapping:
+                    enum: ['first', 'last']
+                    desc: |
+                          If first, will run this command once when the first image is flashed, if
+                          last, will runt his command once when the final image is flashed.
```
Fixed

nordicjm committed:
doc: build: flashing: Add documentation

Adds documentation on the new flashing configuration system

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
doc: release: 3.7: Add SoC/board flashing configuration note

Adds a note that a new flashing configuration system has been
added for boards and socs

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

57300 commented on scripts/west_commands/run_common.py:
```diff
@@ -165,21 +195,101 @@ def do_run_common(command, user_args, user_runner_args, domains=None):
+                build_dir = d.build_dir
+
+            cache = load_cmake_cache(build_dir, user_args)
+            board = cache['CACHED_BOARD']
```
Just remembered that `CACHED_BOARD` could be abbreviated like `nrf5340dk//cpunet`, where the regexes won't work.
Although you have to deliberately put e.g. `-Dremote_BOARD=nrf5340dk//cpunet` and run `--recover` to notice this.
Still, maybe `CONFIG_BOARD_TARGET` is needed here instead.

nordicjm committed:
scripts: list_hardware: Add check for soc.yml runner validity

Adds checking that qualifiers listed in a soc.yml file are valid
for the socs and cpuclusters defined in that file

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

carlescufi commented on doc/build/flashing/configuration.rst:
```diff
@@ -0,0 +1,92 @@
+
+.. code-block:: yaml
+
+  runners:
+    run_once:
+      - '--recover':
+          run: first
+          runners:
+            - nrfjprog
+          groups:
+            - qualifiers:
+                - (.+)/nrf5340/cpunet
+                - (.+)/nrf5340/cpuapp
+                - (.+)/nrf5340/cpuapp/ns
+      - '--reset':
+          run: last
+          runners:
+            - nrfjprog
+            - jlink
+          groups:
+            - qualifiers:
+                - (.+)/nrf5340/cpunet
+            - qualifiers:
+                - (.+)/nrf5340/cpuapp
+                - (.+)/nrf5340/cpuapp/ns
```
I discussed this internally with @nordicjm and @tejlmand but I will post my opinion on it here.
My problem with this approach is that you are essentially using the following format, translating the `runners` node into pseudo-Python:

`{run_once: [{'--recover': {run: ... , runners: ...}}, {'--reset': {}} ]}`

so it's a map, and the first key is `run_once`. Then, the value corresponding to the key `run_once` is what I was an array of dictionaries, where each of these dictionaries has a single key which is the name of the command (`--recover`, `--reset`). The value of that key then is in turn a map with a bunch of keys (`run`, `runners`, `groups`, ...).
I personally dislike this approach because you end up with an array of single-key/value dictionaries. Your rationale for this choice is that snippets currently use this format, so it's a matter of consistency. But I partially disagree with that statement, because the snippet .yaml files use a significantly different approach: instead of having an array of single-key/value dictionaries, snippets use a single dictionary in which each key is a regular expression. I think that the snippet approach is valid, because it uses the dictionary data structure in a way that (to me at least) makes sense: filtering keys and then looking up the value. You pointed out that the exact same approach that snippets use is not possible here, due to the fact that, in the `run_once` case, the commands can appear more than once, so they cannot be dictionary keys, hence the use of an array.
Now, in that case, and since we have an array anyway, I do prefer @tejlmand's proposal of an array of dictionaries, where each dictionary has multiple key/value pairs and one of which is then `command:` (key) `--recover` (value). Because I think it makes more sense to use an array in this fashion, and also reduces the nesting by one level. With this approach you get, again in Python pseudo-code:

`{run_once: [{command: '--recover', run: ... , runners: ...}, {command: '--reset', run: ..., runners: ...} ]}`

To make my point about the nesting levels a bit more clear, I'd like to compare accessing an item in both proposals. I will compare the `run` dict item to `first`:

In your initial proposal: 
`run_once[i]['--recover'][run] == first`
where `i` is the item index in the `run_once` sequence that has `--recover` as its dict key

Alternative approach:
`run_once[i][run] == first`
where `i` is the item index in the `run_once` sequence that has `--recover` as the value of its command key

fabiobaltieri replied:
Well the whole format sounds a bit weird, the whole "run once" -> "first/last" thing. Would it make sense to instead have something like a "pre_flash {command, runner...}" and "post_flash {command, runner...}"? Anything to make this a bit more compact as well.

carlescufi replied:
I don't think so @fabiobaltieri. The nature of the combinations needed requires a system like this, because you need to be able to specify even multiple times a single command to be run once. pre_flash and post_flash just wouldn't be flexible enough :(

carlescufi replied:
I've now seen the latest proposal, which effectively translates to:

`{run_once: {'--recover': [{run: ... , runners: ...}], '--reset': [{run: ..., runners: ...}, {run: ..., runners: ...}] }}`

This indeed uses a single dict with the different commands as the keys, which is much closer to what snippets do, and much cleaner from my PoV.

nordicjm commented:
@dleach02 not sure if this feature would be useful for your platforms (e.g. tf-m?), would you be able to review?

tejlmand reviewed (dismissed):
Still disapprove the approach of placing a value as the key.

This example:
```
  runners:
    run_once:
      '--recover':
        - run: first
          runners:
            - nrfjprog
          groups:
            - qualifiers:
                - ...
      '--reset':
        - run: last
          runners:
            - ...
```
Ref: https://github.com/zephyrproject-rtos/zephyr/blob/9c570c0c4599e00fd689648d2b2c8dcefeb68fa2/doc/build/flashing/configuration.rst?plain=1#L48

together with this code:
https://github.com/zephyrproject-rtos/zephyr/blob/9c570c0c4599e00fd689648d2b2c8dcefeb68fa2/scripts/list_hardware.py#L46-L50

cleary shows that:
`groups` is a proper sequence in the yaml, and treated as one in the code.
`qualifiers` is a proper sequence in the yaml, and treated as one in the code.
`grp` (`command`) is not a sequence in the yaml, but treated as one in the code.

only difference between the `grp` list in code is that due to it being described using the key field, then each entry must be unique, so it's more akin to a set than a list.
But an opaque set (list) because the lack of a `-` in the updated proposal.

So I still argue that a proper yaml scheme of `Key: value`, should be used, for example `command: '--reset'`

nordicjm commented:
> Still disapprove the approach of placing a value as the key.
> 
> This example:
> 
> ```
>   runners:
>     run_once:
>       '--recover':
>         - run: first
>           runners:
>             - nrfjprog
>           groups:
>             - qualifiers:
>                 - ...
>       '--reset':
>         - run: last
>           runners:
>             - ...
> ```
> 
> Ref:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/9c570c0c4599e00fd689648d2b2c8dcefeb68fa2/doc/build/flashing/configuration.rst?plain=1#L48
> 
> together with this code:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/9c570c0c4599e00fd689648d2b2c8dcefeb68fa2/scripts/list_hardware.py#L46-L50
> 
> cleary shows that: `groups` is a proper sequence in the yaml, and treated as one in the code. `qualifiers` is a proper sequence in the yaml, and treated as one in the code. `grp` (`command`) is not a sequence in the yaml, but treated as one in the code.
> 
> only difference between the `grp` list in code is that due to it being described using the key field, then each entry must be unique, so it's more akin to a set than a list. But an opaque set (list) because the lack of a `-` in the updated proposal.
> 
> So I still argue that a proper yaml scheme of `Key: value`, should be used, for example `command: '--reset'`

The above is incorrect, the structure here is compact and minimal, it has a map for the commands because there are a map NOT a list as is claimed above, given the following example:
```
   "runners": {
       "run_once": {
           "--recover": [
               {
                   "runners": [
                       "nrfjprog"
                   ],
                   "run": "first",
                   "groups": [
                       {
                           "qualifiers": [
                               "nrf51([0-9]{3})((.+)?)"
                           ]
                       },
                       {
                           "qualifiers": [
                               "nrf52([0-9]{3})((.+)?)"
                           ]
                       },
                       {
                           "qualifiers": [
                               "nrf5340/cpunet",
                               "nrf5340/cpuapp",
                               "nrf5340/cpuapp/ns"
                           ]
                       },
                       {
                           "qualifiers": [
                               "nrf9160",
                               "nrf9160/ns"
                           ]
                       }
                   ]
               }
           ],
           "--erase": [
               {
...
```

There can be one command per set of configuration, for that command there is a list whereby each list corresponds to one or more flash runners (or all of them if one has the `all` value). The method posted in the above would be like having a full database setup with a foreign key to a table named `commands` then for some reason having each foreign key have the text `command` prefixed to it, so instead of concise numerical foreign keys like `2`, `7`, `9`, etc. you would have `command2`, `command7`, `command9`, it adds nothing, wastes space, slows everything down and looks an eyesore because you're using something that is implicitly obvious, it's completely redundant.

erwango commented on doc/build/flashing/configuration.rst:
```diff
@@ -0,0 +1,105 @@
+*************
+
+Configuration is applied by using a ``runners`` map with a ``run_once`` child, from here is a
+dictionary of commands as they would be provided to the flash runner e.g. ``--reset`` following by
```
>  from here is a

unclear

nordicjm replied:
Have reworded

erwango commented on doc/build/flashing/configuration.rst:
```diff
@@ -0,0 +1,105 @@
+Configuration is applied by using a ``runners`` map with a ``run_once`` child, from here is a
+dictionary of commands as they would be provided to the flash runner e.g. ``--reset`` following by
+a list which specifies the settings for each of these commands (these are grouped by flash runner,
+and by qualifiers/boards). Commands have associated runners that they apply to using a ``runners``
```
s/to using/to use

nordicjm replied:
to using is correct here

erwango commented on doc/build/flashing/configuration.rst:
```diff
@@ -0,0 +1,105 @@
+a list which specifies the settings for each of these commands (these are grouped by flash runner,
+and by qualifiers/boards). Commands have associated runners that they apply to using a ``runners``
+list value, this can contain ``all`` if it applies to all runners, otherwise must contain each
+runner that it applies to using the runner-specific name. Groups of board targets can be specified
+using the ``groups`` key which has a list of board target sets - board targets are regular
+expression matches, for ``soc.yml`` files each set of board targets must be in a ``qualifiers`` key
+(only regular expression matches for board qualifiers are allowed, the board name must be omitted
+from these entries) and for ``board.yml`` files each set of board targets must be in a ``boards``
+key, these are lists containing the matches which form a singular group. A final parameter ``run``
```
This long sentence should be slplit

nordicjm replied:
Split up

erwango commented on doc/build/flashing/configuration.rst:
```diff
@@ -0,0 +1,105 @@
+Usage
+*****
+
+Commands that are supported by flash runners can be used as normal, when flashing a sysbuild
+project with multiple images, the flashing configuration run once settings will be applied.
```
Sentence to be reviewed: " the flashing configuration run once settings will be applied." I don't understand

nordicjm replied:
Reworded

JordanYates commented on scripts/west_commands/run_common.py:
```diff
@@ -201,6 +320,93 @@ def do_run_common_image(command, user_args, user_runner_args, build_dir=None):
+    # and if so, remove them for all but the first iteration of the flash
+    # runner per unique board name.
+    if len(used_cmds) > 0 and len(runner_args) > 0:
+        i = len(runner_args) - 1
```
Given `i` only ever goes down by 1 each iteration, I think the following is much easier to read:
`for i in reversed(range(len(runner_args))):`

JordanYates reviewed (commented):
My general feedback as requested by @carlescufi.
Focusing purely on the yaml format, the python equivalent (I'm scared of messing up the yaml indentation in a proposal) I would use is:
```
"runners": {
  "run_first": [
    {"command": "--recover", "runners": [], ...}
  ],
  "run_last":  [
    {"command": "--reset", "runners": [], ...},
    {"command": "--something", "runners": [], ...}
  ]
}
```
I don't see the value of the `run_once` level, and suspect that having `run_first` and `run_last` as the top level keys would make the flashing logic simpler as you don't need to inspect as deeply to find the `run` key.

If this is the first/last image, look under `run_first/run_last`. If options exist, scan to see if there is a matching runner & qualifiers. If a match is found, apply `command`.

As from my understanding you can have an arbitrary number of options under `run_once` (or `run_first/last`, it makes sense to me for that structure to be a list, not a dictionary keyed by the command.

nordicjm commented:
> My general feedback as requested by @carlescufi. Focusing purely on the yaml format, the python equivalent (I'm scared of messing up the yaml indentation in a proposal) I would use is:
> 
> ```
> "runners": {
>   "run_first": [
>     {"command": "--recover", "runners": [], ...}
>   ],
>   "run_last":  [
>     {"command": "--reset", "runners": [], ...},
>     {"command": "--something", "runners": [], ...}
>   ]
> }
> ```
> 
> I don't see the value of the `run_once` level, and suspect that having `run_first` and `run_last` as the top level keys would make the flashing logic simpler as you don't need to inspect as deeply to find the `run` key.
> 
> If this is the first/last image, look under `run_first/run_last`. If options exist, scan to see if there is a matching runner & qualifiers. If a match is found, apply `command`.
> 
> As from my understanding you can have an arbitrary number of options under `run_once` (or `run_first/last`, it makes sense to me for that structure to be a list, not a dictionary keyed by the command.

The python code would need to merge these together so that's why in this PR there is a single key and one of the values is if it runs first or last, having it as 2 lists would mean that the code creating the objects in python would need to be duplicated for the other one, this is how the python code constructs it:
```
                for cmd in data_yaml['runners']['run_once']:
                    for data in data_yaml['runners']['run_once'][cmd]:
                        for group in data['groups']:
                            run_first = bool(data['run'] == 'first')
                            if group_type == 'qualifiers':
                                targets = []
                                for target in group[group_type]:
                                    # For SoC-based qualifiers, prepend to the beginning of the
                                    # match to allow for matching any board name
                                    targets.append('([^/]+)/' + target)
                            else:
                                targets = group[group_type]

                            used_cmds.append(UsedFlashCommand(cmd, targets, data['runners'], run_first))
```
where `run_first` is a bool, it is just stored inside a python list and accessed from that list in the code below it which isn't very complex.

> If this is the first/last image, look under run_first/run_last. If options exist, scan to see if there is a matching runner & qualifiers. If a match is found, apply command.

Can't be done like that because it is grouped for a command, for a set of flash runners, for a set of boards, so you might have a board with 2 SoCs on it e.g. nrf9160 and nrf52840, you can't just do something on the first and last images because the flash order might be nrf9160 mcuboot -> nrf52840 mcuboot -> nrf9160 application -> nrf52840 application, the groupings would need to apply to all nrf52840 images and all nrf9160 images separately, so if someone uses --recover then it would be nrf9160 mcuboot (with --recover) -> nrf52840 mcuboot (with --recover) -> nrf9160 application (with --reset) -> nrf52840 application (with --reset)

JordanYates commented:
> Can't be done like that because it is grouped for a command, for a set of flash runners, for a set of boards, so you might have a board with 2 SoCs on it e.g. nrf9160 and nrf52840, you can't just do something on the first and last images because the flash order might be nrf9160 mcuboot -> nrf52840 mcuboot -> nrf9160 application -> nrf52840 application, the groupings would need to apply to all nrf52840 images and all nrf9160 images separately, so if someone uses --recover then it would be nrf9160 mcuboot (with --recover) -> nrf52840 mcuboot (with --recover) -> nrf9160 application (with --reset) -> nrf52840 application (with --reset)

How does `west flash` know which JTAGs to use for which cores in this situation?
Unless there is some magic parallel SWD feature I am unaware of?

nordicjm commented:
> > Can't be done like that because it is grouped for a command, for a set of flash runners, for a set of boards, so you might have a board with 2 SoCs on it e.g. nrf9160 and nrf52840, you can't just do something on the first and last images because the flash order might be nrf9160 mcuboot -> nrf52840 mcuboot -> nrf9160 application -> nrf52840 application, the groupings would need to apply to all nrf52840 images and all nrf9160 images separately, so if someone uses --recover then it would be nrf9160 mcuboot (with --recover) -> nrf52840 mcuboot (with --recover) -> nrf9160 application (with --reset) -> nrf52840 application (with --reset)
> 
> How does `west flash` know which JTAGs to use for which cores in this situation? Unless there is some magic parallel SWD feature I am unaware of?

That's a good question, not something I've tried, there is a `-snr` flash runner option for specifying which one to use, not sure if that can be set per image in a file @PerMac any ideas? I guess nrf9160 and nrf52840 are probably bad examples to use, so you can replace them with nrf5340 application core and nrf5340 network core

PerMac commented:
> > > Can't be done like that because it is grouped for a command, for a set of flash runners, for a set of boards, so you might have a board with 2 SoCs on it e.g. nrf9160 and nrf52840, you can't just do something on the first and last images because the flash order might be nrf9160 mcuboot -> nrf52840 mcuboot -> nrf9160 application -> nrf52840 application, the groupings would need to apply to all nrf52840 images and all nrf9160 images separately, so if someone uses --recover then it would be nrf9160 mcuboot (with --recover) -> nrf52840 mcuboot (with --recover) -> nrf9160 application (with --reset) -> nrf52840 application (with --reset)
> > 
> > How does `west flash` know which JTAGs to use for which cores in this situation? Unless there is some magic parallel SWD feature I am unaware of?
> 
> That's a good question, not something I've tried, there is a `-snr` flash runner option for specifying which one to use, not sure if that can be set per image in a file @PerMac any ideas? I guess nrf9160 and nrf52840 are probably bad examples to use, so you can replace them with nrf5340 application core and nrf5340 network core

I won't be able to look at this for 2 weeks. Maybe @gchwier ?

nordicjm commented:
@fabiobaltieri @JordanYates is the PR acceptable? If so, can you approve please?

fabiobaltieri commented:
> @fabiobaltieri @JordanYates is the PR acceptable? If so, can you approve please?

To be totally honest I don't love it, it seems fairly complex, but I also don't really have any alternative to suggest so fine by me.

carlescufi commented:
@nordicjm please rebase

carlescufi reviewed (approved):
As described [here](https://github.com/zephyrproject-rtos/zephyr/pull/69748#discussion_r1551767151) and [here](https://github.com/zephyrproject-rtos/zephyr/pull/69748#discussion_r1557624048) I am not a huge fan of the yaml format, but I see no reason not to move forward with this at this point.

carlescufi commented:
Removing assignee since @tejlmand is opting out of this particular review.
